### PR TITLE
worker/uniter: Remove temp download files

### PR DESCRIPTION
**Pull request by mjs**
The temporary files created during charm downloads were being left behind if an error occured. In some cases (a repeating permanent error) this would lead to disk space exhaustion.

The uniter will now also clear any temporary charm download files for the unit it manages when it starts. Although this shouldn't happen now, previous versions of Juju could leave a significant amount of data behind from failed download attempts.

This completes the fixes for https://bugs.launchpad.net/juju/+bug/1626304.

### CODE REVIEW

File: worker/uniter/charm/bundles.go [hunk 1]
```diff
@@ -122,8 +133,29 @@ func (d *BundlesDir) bundleURLPath(url *charm.URL) string {
...
+	entries, err := ioutil.ReadDir(downloadDir)
+	if os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return
+	}
+	for _, entry := range entries {
+		path := filepath.Join(downloadDir, entry.Name())
+		if err = os.RemoveAll(path); err != nil {
+			return
```
Comment (mjs) [inline worker/uniter/charm/bundles.go]: Thinking about it, this should probably keep going with other entries even if one fails to remove.

File: worker/uniter/uniter.go [hunk 2]
```diff
@@ -236,6 +236,9 @@ func (u *Uniter) init(unitTag names.UnitTag) (err error) {
 	u.storage = storageAttachments
 	u.addCleanup(storageAttachments.Stop)
 
+	if err := charm.ClearDownloads(u.paths.State.BundlesDir); err != nil {
+		return errors.Annotatef(err, "cannot clear downloads")
```
Comment (mjs) [inline worker/uniter/uniter.go]: This probably shouldn't be fatal should it?... Might be best to just warn instead.

File: worker/uniter/charm/bundles.go (hunk 1 above)
Comment (axw) [inline worker/uniter/charm/bundles.go]: +1, though I'm not sure how helpful it'll be in practice. Assuming it's a transient error, the worker should restart and eventually clean them all up.

File: worker/uniter/charm/bundles.go [hunk 3]
```diff
@@ -131,8 +133,29 @@ func (d *BundlesDir) bundleURLPath(url *charm.URL) string {
 	return path.Join(d.path, charm.Quote(url.String()))
 }
 
+// ClearDownloads removes any entries in the temporary bundle download
+// directory. It is intended to be called on uniter startup.
+func ClearDownloads(bundlesDir string) (err error) {
+	defer errors.DeferredAnnotatef(&err, "unable to clear bundle downloads")
+
+	downloadDir := downloadsPath(bundlesDir)
```
Comment (axw) [inline worker/uniter/charm/bundles.go]: Maybe just `os.RemoveAll(downloadDir)`? The downloader will recreate it as necessary.

File: worker/uniter/uniter.go (hunk 2 above)
Comment (axw) [inline worker/uniter/uniter.go]: I'd prefer if we would just bail, but let's warn for now to be on the safe side. Something external could block the deletion of the directory/files by holding onto them. Shouldn't happen, but it's possible, and I don't think we should risk that in a maintenance branch.

File: worker/uniter/charm/bundles.go (hunk 3 above)
Comment (mjs) [inline worker/uniter/charm/bundles.go]: Done

**Comment by mjs** [comment]
$$merge$$

**Pull request merged**